cf-serverd/server_tls.c

```diff
@@ -90,87 +115,91 @@ bool ServerTLSInitialize()
         "Setting cipher list for incoming TLS connections to: %s",
         cipher_list);
 
-    ret = SSL_CTX_set_cipher_list(SSLSERVERCONTEXT, cipher_list);
+    ret = SSL_CTX_set_cipher_list(*ssl_ctx, cipher_list);
     if (ret != 1)
     {
         Log(LOG_LEVEL_ERR,
             "No valid ciphers in cipher list: %s",
             cipher_list);
-        goto err2;
+        goto err;
     }
 
-    if (PRIVKEY == NULL || PUBKEY == NULL)
-    {
-        Log(LOG_LEVEL_ERR, "No public/private key pair is loaded,"
-            " please create one using cf-key");
-        goto err2;
-    }
-
-    assert(SSLSERVERCERT == NULL);
     /* Create cert into memory and load it into SSL context. */
-    SSLSERVERCERT = TLSGenerateCertFromPrivKey(PRIVKEY);
-    if (SSLSERVERCERT == NULL)
+    X509 *cert = TLSGenerateCertFromPrivKey(priv_key);
+    if (cert == NULL)
     {
         Log(LOG_LEVEL_ERR,
             "Failed to generate in-memory certificate from private key");
-        goto err2;
+        goto err;
     }
 
-    SSL_CTX_use_certificate(SSLSERVERCONTEXT, SSLSERVERCERT);
+    SSL_CTX_use_certificate(*ssl_ctx, cert);
+    X509_free(cert);
```

Comment:
Did you find that the cert is not needed after SSL_CTX_use_certificate? Previously the code waited until later to free the cert. I looked into the openssl code a bit and didn't see something that clearly said you could free the cert.

Reply:
I did the same with the same result. So I tried it. And it seems to be working just fine. There's no way to get the cert back from the SSL_CTX so I think it must somehow load the cert into it's own internal state.